src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -290,6 +290,27 @@ public HubConnectionState getConnectionState() {
         return hubConnectionState;
     }
 
+    // For testing only
+    String getBaseUrl() {
+        return this.baseUrl;
+    }
+
+    /**
+     * Sets a new url for the HubConnection.
+     * @param url The url to connect to.
+     */
+    public void setBaseUrl(String url) {
+        if (url == null || url.isEmpty()) {
+            throw new IllegalArgumentException("The HubConnection url must be a valid url.");
+        }
+
+        if (hubConnectionState != HubConnectionState.DISCONNECTED) {
+            throw new IllegalStateException("The HubConnection must be in the disconnected state to change the url.");
```

Comment:
Are we guaranteed that the connection will be in the `DISCONNECTED` state during the onClosed callback? I'd like to see a test confirming that works (changing the URL in the `onClosed` callback, and doing so **without** waiting for `.stop` to complete, to make sure the asynchrony is working right).